libs/vkd3d/swapchain_new.c

```diff
@@ -0,0 +1,1627 @@
+    return false;
+}
+
+static UINT STDMETHODCALLTYPE dxgi_vk_swap_chain_CheckColorSpaceSupport(IDXGIVkSwapChain *iface, DXGI_COLOR_SPACE_TYPE ColorSpace)
+{
+    struct dxgi_vk_swap_chain *chain = impl_from_IDXGIVkSwapChain(iface);
+    UINT support_flags = 0;
+    if (dxgi_vk_swap_chain_supports_color_space(chain, ColorSpace))
+        support_flags |= DXGI_SWAP_CHAIN_COLOR_SPACE_SUPPORT_FLAG_PRESENT;
+    return support_flags;
+}
+
+static HRESULT STDMETHODCALLTYPE dxgi_vk_swap_chain_SetColorSpace(IDXGIVkSwapChain *iface, DXGI_COLOR_SPACE_TYPE ColorSpace)
+{
+    struct dxgi_vk_swap_chain *chain = impl_from_IDXGIVkSwapChain(iface);
+    if (!dxgi_vk_swap_chain_supports_color_space(chain, ColorSpace))
+        return E_INVALIDARG;
+
+    chain->user.dxgi_color_space_type = ColorSpace;
+    chain->user.modifies_hdr_metadata = true;
+
+    return S_OK;
+}
+
+static HRESULT STDMETHODCALLTYPE dxgi_vk_swap_chain_SetHDRMetaData(IDXGIVkSwapChain *iface, const DXGI_VK_HDR_METADATA *pMetaData)
+{
+    struct dxgi_vk_swap_chain *chain = impl_from_IDXGIVkSwapChain(iface);
+    chain->user.dxgi_hdr_metadata = *pMetaData;
+    chain->user.modifies_hdr_metadata = true;
+    return S_OK;
+}
+
+static CONST_VTBL struct IDXGIVkSwapChainVtbl dxgi_vk_swap_chain_vtbl =
+{
+    /* IUnknown methods */
+    dxgi_vk_swap_chain_QueryInterface,
+    dxgi_vk_swap_chain_AddRef,
+    dxgi_vk_swap_chain_Release,
+
+    /* IDXGIVkSwapChain methods */
+    dxgi_vk_swap_chain_GetDesc,
+    dxgi_vk_swap_chain_GetAdapter,
+    dxgi_vk_swap_chain_GetDevice,
+    dxgi_vk_swap_chain_GetImage,
+    dxgi_vk_swap_chain_GetImageIndex,
+    dxgi_vk_swap_chain_GetFrameLatency,
+    dxgi_vk_swap_chain_GetFrameLatencyEvent,
+    dxgi_vk_swap_chain_ChangeProperties,
+    dxgi_vk_swap_chain_SetPresentRegion,
+    dxgi_vk_swap_chain_SetGammaControl,
+    dxgi_vk_swap_chain_SetFrameLatency,
+    dxgi_vk_swap_chain_Present,
+    dxgi_vk_swap_chain_CheckColorSpaceSupport,
+    dxgi_vk_swap_chain_SetColorSpace,
+    dxgi_vk_swap_chain_SetHDRMetaData,
+};
+
+static HRESULT dxgi_vk_swap_chain_create_surface(struct dxgi_vk_swap_chain *chain, HWND hwnd)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &chain->queue->device->vk_procs;
+#ifdef VK_KHR_win32_surface
+    VkWin32SurfaceCreateInfoKHR create_info;
+#endif
+    VkPhysicalDevice vk_physical_device;
+    VkInstance vk_instance;
+    VkBool32 supported;
+    VkResult vr;
+
+    vk_instance = chain->queue->device->vkd3d_instance->vk_instance;
+    vk_physical_device = chain->queue->device->vk_physical_device;
+
+#ifdef VK_KHR_win32_surface
+    create_info.sType = VK_STRUCTURE_TYPE_WIN32_SURFACE_CREATE_INFO_KHR;
+    create_info.pNext = NULL;
+    create_info.hwnd = hwnd;
+    create_info.hinstance = GetModuleHandleA("d3d12.dll");
+    create_info.flags = 0;
+    vr = VK_CALL(vkCreateWin32SurfaceKHR(vk_instance, &create_info, NULL, &chain->vk_surface));
+#else
+    /* TODO: With dxvk-native integration, we can modify this as needed. */
```

Comment:
I think we should let DXVK create the surface -- even for non-native builds.

There are two options here:
 - DXVK DXGI owns the surface but we need to share a VkInstance
 - We own the surface and ask for it to be made with a specific instance

 I am kind of in favour of the former as it seems cleaner overall.

I dont want to duplicate any logic here as I eventually I do want to add non-SDL WSI to DXVK and I don't want to have copy pasta when we can solve this elegantly.

Reply:
Sharing VkInstance could get annoying due to debug extensions, DXVK would have to know about those. Having DXVK create the surface would be cleaner.

Reply:
We're using the new DXVK SurfaceFactory interface, so this should be resolved now.